crates/transport-http/src/layers/auth.rs

```diff
@@ -0,0 +1,158 @@
+};
+use alloy_rpc_types_engine::{Claims, JwtSecret};
+use alloy_transport::{TransportError, TransportErrorKind};
+use hyper::header::HeaderValue;
+use jsonwebtoken::{decode, get_current_timestamp, DecodingKey, Validation};
+use std::{
+    future::Future,
+    pin::Pin,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use tower::{Layer, Service};
+
+/// The [`AuthLayer`] is a that validates whether the bearer token in the request is valid or not.
+/// If invalid, it generates a valid token from the provided [`JwtSecret`] and inserts it into the
+/// request.
+///
+/// This layer also inserts the [`AUTHORIZATION`] header into the request with a valid token if its
+/// not already in the request.
+#[derive(Clone, Debug)]
+pub struct AuthLayer {
+    secret: JwtSecret,
+    latency_buffer: u64,
+}
+
+impl AuthLayer {
+    /// Create a new [`AuthLayer`].
+    pub const fn new(secret: JwtSecret) -> Self {
+        Self { secret, latency_buffer: 5000 }
+    }
+
+    /// We use this buffer to perfom an extra check on the `iat` field to prevent sending any
+    /// requests with tokens that are valid now but may not be upon reaching the server.
+    ///
+    /// In milliseconds. Default is 5s.
+    pub const fn with_latency_buffer(self, latency_buffer: u64) -> Self {
+        Self { latency_buffer, ..self }
+    }
+}
+
+impl<S> Layer<S> for AuthLayer {
+    type Service = AuthService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        AuthService::new(inner, self.secret, self.latency_buffer)
+    }
+}
+
+/// A service that checks the jwt token in a request is valid. If invalid, it refreshes the token
+/// using the provided [`JwtSecret`] and inserts it into the request.
+#[derive(Clone, Debug)]
+pub struct AuthService<S> {
+    inner: S,
+    secret: JwtSecret,
+    /// In milliseconds.
+    latency_buffer: u64,
+}
+
+impl<S> AuthService<S> {
+    /// Create a new [`AuthService`] with the given inner service.
+    pub const fn new(inner: S, secret: JwtSecret, latency_buffer: u64) -> Self {
+        Self { inner, secret, latency_buffer }
+    }
+
+    /// Validate the token in the request headers.
+    ///
+    /// Returns `true` if the token is valid and `iat` is beyond the grace buffer.
+    pub fn validate(&self, headers: &HeaderMap) -> bool {
+        get_bearer_token(headers).map_or(false, |token| {
+            let is_valid = self.secret.validate(&token).ok().and_then(|_| {
+                let validation = Validation::new(jsonwebtoken::Algorithm::HS256);
+                let decoding_key = DecodingKey::from_secret(self.secret.as_bytes());
+                decode::<Claims>(token.as_str(), &decoding_key, &validation).ok().and_then(|data| {
+                    let curr_secs = get_current_timestamp();
+                    if data.claims.iat.abs_diff(curr_secs) * 1000 <= self.latency_buffer {
+                        None
+                    } else {
+                        Some(())
+                    }
+                })
```

Comment:
all of this is a bit costly and kinda redundant because all we do here is validating our own claims, if we track the most recent claim instead we can make this easier